### ISSUE TITLE
Opinion: Storing scores inside each Metric might not be ideal

### ISSUE THREAD

**Issue by salomaestro** [SFI-Visual-Intelligence/Collaborative-Coding-Exam#84]
TLDR; I propose that we reconsider the approach of storing and returning scores directly within each metric by using `MyMetric.__returnmetric__()`. Instead, returning to the previous design may offer a simpler and more intuitive workflow.

---

While this is just a small exam, and does not really matter, I think it's worth discussing our current metric design.

The existing `MetricWrapper` class is helpful, but I never quite understood the issue with allowing it to handle the accumulation and storage of metrics itself—this seemed more intuitive to me.

Under the new approach, each individual metric is responsible for storing its own score. However, by doing so, we lose the straightforward testing approach that PyTorch encourages with `nn.Module`. Previously, all you had to do was implement `Module.forward()` and call something like:

```python
precision = Precision(num_classes=2)
score = precision(y_true, logits)
```

With the current design, you instead need to do:

```python
precision = Precision(num_classes=2)
precision(y_true, logits)
score = precision.__returnmetric__()
```

This additional step might confuse new users or those unfamiliar with our design choices—requiring them to consult the documentation to understand our custom metric workflow.

I’d love to hear everyone’s thoughts on whether reverting to the old design could simplify usage and testing. Of course, I’m open to being wrong and eager to learn if there are advantages to the current implementation that I’m missing.

**Comment by Seilmast** [comment]
The main idea behind storing values in individual classes are to 
1. Allow us to work on our separate tasks. 
2. Unsure if storing intermediary values is actually possible for metrics such as F1

Like, ideally we'd just have a single class which logs for example logit values and contains a confusion matrix, then returns all relevant metrics once called upon. However, our exam doesn't really allow for that.
We could probably redesign the class, but in my opinion, I just want to finish the exam and not have to wait for several more changes to everyones metrics.

**Comment by Johanmkr** [comment]
Personally, I don't mind as I am not used to any particular way of organizing this stuff. I suggest those with strong opinions to make their points at the meeting today (20th) and then we can settle on one method that everyone follows, so we can proceed on the remaining tasks.

**Comment by hzavadil98** [comment]
Resolved - keeping the metric based storing of values to compute the metrics over the whole dataset.
